### PR TITLE
Add dialyzer to CI

### LINKED ISSUE #25: dialyzer in ex_plasma as a circleci job

### PR DESCRIPTION
**Pull request by achiurizo**
closes #25 

## What changed and why?

Adds dialyzer to our CI. See pull request build checks.

## How was this tested?

Dialyzer trying to help me:

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -66,6 +66,37 @@ jobs:
...
+      - image: circleci/elixir:1.9.2
+    environment:
+      MIX_ENV=dev
+    steps:
+      - checkout
+      - run: mix local.rebar --force && mix local.hex --force
+      - restore_cache:
+          key: v1-mix-cache-{{ .Branch }}-{{ checksum "mix.lock" }}
+      - restore_cache:
+          key: v1-dialyzer-deps-cache-{{ .Branch }}-{{ checksum "mix.lock" }}
```
Comment (unnawut) [inline .circleci/config.yml]:
```suggestion
          key: v1-dialyzer-plt-cache-{{ .Branch }}-{{ checksum "mix.lock" }}
```

got confused for a moment whether it's restoring deps or plts

**Comment by unnawut** [comment]
The CI part looks good. The ignores though, I think most if not all can be fixed?

**Comment by achiurizo** [comment]
> The CI part looks good. The ignores though, I think most if not all can be fixed?

@unnawut oh yeah, I'm just terrible with dialyzer so I kicked that can down the road for another PR 😆

**Comment by achiurizo** [comment]
paired with @unnawut and fixed most of the dialyzer errors except the `defprotocol` related ones. Currently ignoring those for now as the solution we have doesn't feel good as well (e.g explicitly setting up implementation for each of the types)

**Pull request merged**